full-text search using Lunr.js

ZoeLeBlanc opened the pull request:
As an alternative to List.js, we could also use Lunr.js which has more robust search capacity. 

This branch has my current attempt to get the two of these working together. It requires pulling down the pre-built indices and corpora, then filtering them for a search term, trying to reuse some of the existing UX by having the search results appear where the lesson abstract usually does. 

One issue I'm having is that I'm not getting the right position for the search terms in the documents. @mdlincoln I'm using your code from the gist so if you wouldn't mind taking a look I would really appreciate it!

closes #1018 

### Checklist

- [x] Assign yourself in the "Assignees" menu
- [x] Assign at least one individual or team to "Reviewers"
  - [x] if the text needs to be translated, assign the relevant language team(s) as "Reviewers" and tag both the team as well as the managing edtor in your PR. Please follow the [translation request guidelines](https://github.com/programminghistorian/jekyll/wiki/Requesting-Translation-Guidelines) when writing your PR description
- [x] Add the appropriate "Label"
- [x] [Ensure the Travis CI checks pass](https://github.com/programminghistorian/jekyll/wiki/Making-Technical-Contributions#4-check-your-pr-status)
- [x] [Check the live preview of your PR on Netlify](https://github.com/programminghistorian/jekyll/wiki/Making-Technical-Contributions#5-preview-how-your-pr-looks-when-built-into-html)
- [x] If this PR closes an open issue, add the phrase `Closes #ISSUENUMBER` to the description above

*If you are having difficulty fixing Travis errors, first consult <https://github.com/programminghistorian/jekyll/wiki/Making-Technical-Contributions> carefully, especially ["Common Travis Errors"](https://github.com/programminghistorian/jekyll/wiki/Making-Technical-Contributions#common-travis-errors). Then contact the technical team if you need further help.*

Commit:
maybe solved searching??

Commit:
Added comments for search

Commit:
attempt at a lunr search

Commit:
Merge branch 'gh-pages' of github.com:programminghistorian/jekyll into search-lunr-1018

Commit:
removed wrap terms func

mdlincoln commented:
I think you're getting the right search position, but applying it to the wrong documents! E.g. if I search `georeferencing` it doesn't bring up the georeferencing lesson at all, but three other ones that don't even mention the term. 

There's an indexing issue somewhere in your code, where you're picking the wrong lessons from the search results and then trying to applying the highlighting to the text bodies and thus getting bizarre results - haven't pinpointed it yet, but that's where I'd look.

mdlincoln commented on js/lessonfilter.js:
```diff
@@ -60,13 +99,57 @@ function wireButtons() {
+  // Get search indices and corpuses. Right now only trying on English
+  let idx;
+  let corpus;
+  $.getJSON("https://raw.githubusercontent.com/programminghistorian/search-index/master/indices/indexEN.json").done(response => {
```
small thing, but pointing to the published github-pages version of this json at https://programminghistorian.github.io/search-index/indices/indexEN.json gives us the benefits of GH Pages' CDN caching. Always prefer that to pointing to raw.githubusercontent.com

ZoeLeBlanc commented:
Thanks for figuring this out @mdlincoln !!! Definitely appreciate the help!! I wonder if it the search corpus and index are mismatched somehow 🤔 

Otherwise it would likely be the results from the idx.search or where I'm creating the var docs 🕵️‍♀️

mdlincoln commented:
When I inspect the results from `idx.search(searchString);` and compare to the ids in `search.json` they are definitely correct! So at least that part is working.

Where is the portion of the code where you try to match that `id` number to the entries in the `<ul>` on the lessons page? Because those integer IDs only exist rn when creating the search.json file 🤔

mdlincoln commented:
huh, so in the `corpus` object you create, the `id` value is totally wrong, but the index location in the `corpus` object is correct?

e.g. at position 34 in `corpus`, the item has `id: 84`. A search for `georeferencing` returns the id 34, and the doc in position 34 in `corpus` is the right doc, it's just that the id doesn't correspond.

Commit:
fixed lunr ref issue

ZoeLeBlanc commented:
Thanks for the feedback @mdlincoln !! I updated the reference to the search index and I think fixed the issue. You were totally right about the doc.id being the wrong thing. Turns out our Lunr ref was looking for url, so I updated the search-index repo and now everything seems to be working. Gonna add in the logic for multiple languages and then I think we're ready for testing 🎉

Commit:
updated search for multilingual logic

Commit:
trying to fix enter bug

ZoeLeBlanc commented:
So a weird bug that I could use your help with @mdlincoln. If you've noticed sometimes it gives a blank answer or doesn't show the updated search results or doesn't erase them when you empty out the input. 

After lots of testing turns out it does all of that if you click enter at the end of the search query, which I'm pretty sure is an issue with Chrome not updating the new html, but could have to do with List.JS as well...

I've tried using jquery trigger, as well as resizing the window, and using setTimeout to hide and show the search results. So far nothing has worked... any thoughts or ideas of what we should try???

Commit:
trying list.js update function

Commit:
Merge branch 'gh-pages' of github.com:programminghistorian/jekyll into search-lunr-1018

Commit:
fixed earlier bug and got search working with filter logic

Commit:
Merge branch 'gh-pages' of github.com:programminghistorian/jekyll into search-lunr-1018

Commit:
fixed search bugs and got it working with filter logic

Commit:
fixed search bugs and got it working with filter logic

ZoeLeBlanc commented:
Sorry for the very similar commits here, for some reason I wasn't seeing my pushes in the Github UI so overdid it a bit.

Anyways, pretty much have search fully working before tomorrow's meeting 🎉 

You can now click on a topic or activity, search within those lessons, undo your search, and still have those lessons selected. 

I also added logic for a button or enter press which is pretty much fixing the display bug 98% of the time (sometimes you need a second enter press).

Looking forward to hearing how it works for others 😅

mdlincoln commented on js/lessonfilter.js:
```diff
@@ -52,6 +52,50 @@ function applySortFromURI(uri,featureList) {
+  const BUFFER = 30 // Number of characters to show for kwic-results
+  const MAX_KWIC = 3
+  docs.map((doc) => {
+    let elementName = '/'+doc.url.split('/').slice(3, ).join('/')
```
wonder if there's a more reliable way to get the lesson language than parsing the url

Commit:
added search enable button and more comments to code

mdlincoln commented:
- Instead of "enable search lessons?" I might write "Start searching" in the initial search button.
- I type in a search term and click Search Lessons. I then edit the search text and click Search Lessons again, but nothing happens. I would expect that it would execute a new search, but instead I first need to click the "Reset" button
- I type in a search term and click Search Lessons. I then click on one of the faceting buttons above (e.g. Digital Publishing), and the list expands to include all the digital publishing lessons plus the text search lessons. I would expect that it would show the intersect of the two queries.
    - HOWEVER If I do the above and then click Search Lessons again, we get the intersect.
- I love that the search query text entry is automatically capitalized, which is a useful way to suggest that its text insensitive. It IS case-insensitive, right? 
- Is the index creation and index search language-sensitive when it does word stemming? Can we set that up if it's not already? (e.g. in English search, "lesson" will also match "lessons" but in Spanish, "lección" will not match "lecciones"
Questions:

- You'd mentioned we could enable more advanced wildcard searching. Do we think this is really necessary, or overkill? I lean towards thinking it'd be overkill - we don't even have 100 lessons right now.

jenniferisasi commented:
- 'Enable search lessons?' to 'Start searching' or 'Search in lessons' 
- I looked for `text*`, then I do `text` (thinking from a general to specific search) and the search is correct or accurate. Now, upon deleting the work, in this case `text` the result don't clear out. When I click the return key or the 'search lessons' tab it takes back to the previous search results, here `text*`, then another return or search lessons and it clears out to the lesson index. 
- In Spanish I realize it shows some singular/plurals and some verb conjugations in the search but not always (not using the * here). Examples: 
-- valor valores
-- texto textos
-- dato datos
-- escribe escribí 
-- cerrar cerrará
-- guardar guardará (so this indicates I think it takes also whatever follows the characters in the search
- In the same way, it didn't do all the plurals or follow the characters...
-- red, didn't show 'redes' 
- After a few searches for the above experiment, it "run out of searches" in other words, I had to reset to see all lessons (had to click a couple of times)
- It is accent mark sensitive - if missing it won't find the word
- The * search retrieves "all" possible results for the use of a root, ie. 'utiliza*' (verbs + nouns)

I did some searches in French and this is what I found there: 
- 'avancée' retrieves avancée, avancées, avancés.
- 'caractéristique' retrieves the plural 'caractéristiques'
- `l'etu`(missing the 

- It doesn't retrieve a string of characters that doesn't show the word completely? I do 'orig' thinking of original. 
-- origin , retrieves original; original retrieves origin 

- It looks to me that after a few searches the search lesson stops working until you reset.

walshbr commented:
I looked into this, and I agree with @mdlincoln's comments. Particularly re: the caps lock in the search box. I don't have anything else to add! Thanks for working on it.

Commit:
Merge branch 'gh-pages' of github.com:programminghistorian/jekyll into search-lunr-1018

ZoeLeBlanc commented:
Thanks for the feedback everyone! Will be looking at this over the week of May 4 and hopefully won't take too long to debug

Commit:
Merge branch 'gh-pages' of github.com:programminghistorian/jekyll into search-lunr-1018

Commit:
fixed most of the bugs in full text search

Commit:
added in support for multiple language stemmers

Commit:
Merge branch 'gh-pages' of github.com:programminghistorian/jekyll into search-lunr-1018

ZoeLeBlanc commented:
Thanks to everyone for the feedback! I've now gone through and fixed the remaining logic in the lesson filter, figured out that I wasn't properly stemming French and Spanish (so fixed those), and mostly just tried to work through all the bugs below. 

Next steps will involve having second round of testing and then I need to comment and clean up the code (also we could refactor in a few places as  well). 

- [x] ~Instead of "enable search lessons?" I might write "Start searching"  in the initial search button. 'Enable search lessons?' to 'Start searching' or 'Search in lessons'~ Went with Start searching since that seemed to be the preference

- [x] _ ~I type in a search term and click Search Lessons. I then edit the search text and click Search Lessons again, but nothing happens. I would expect that it would execute a new search, but instead I first need to click the "Reset" button~ Seems like it's fixed with the latest changes

- [x] ~I type in a search term and click Search Lessons. I then click on one of the faceting buttons above (e.g. Digital Publishing), and the list expands to include all the digital publishing lessons plus the text search lessons. I would expect that it would show the intersect of the two queries.
 HOWEVER If I do the above and then click Search Lessons again, we get the intersect.~ Gave up my firstborn to get this done, but finally have this working (we should likely DRY up the lessonfilter code at some point the future)

- [x] ~It IS case-insensitive, right? Particularly re: the caps lock in the search box. I don't have anything else to add! Thanks for working on it.~ Double checked and it is case insensitive (but interestingly does not have support for case sensitive)

- [x] ~Is the index creation and index search language-sensitive when it does word stemming?~ 
Turns out it isn't language-sensitive, but lunr does have that option built-in so I've made the necessary changes to the search-index repo and added some js files to this repo. We might want to clean up lesson-index so spanish and french stemmers don't load on every page.

- [x] ~You'd mentioned we could enable more advanced wildcard searching. Do we think this is really necessary, or overkill? I lean towards thinking it'd be overkill - we don't even have 100 lessons right now.~ As far as I can tell wildcard isn't a setting but something that is built-in. If we want to prevent it we would have to add code checking for asterisks in search strings, but honestly few humanists even know the concept so I'm too concerned.

- [x] ~I looked for text*, then I do text (thinking from a general to specific search) and the search is correct or accurate. Now, upon deleting the work, in this case text the result don't clear out. When I click the return key or the 'search lessons' tab it takes back to the previous search results, here text*, then another return or search lessons and it clears out to the lesson index.~ Issue with the search results clearing is now fixed. In terms of results for text* versus text, both turn up almost identical hits because many lessons with texts also have text. A more helpful example is net* vs network, which should show some pretty big variation

- [x] In Spanish I realize it shows some singular/plurals and some verb conjugations in the search but not always (not using the * here). Examples:
-- valor valores
-- texto textos
-- dato datos
-- escribe escribí
-- cerrar cerrará
-- guardar guardará (so this indicates I think it takes also whatever follows the characters in the search
In the same way, it didn't do all the plurals or follow the characters...
-- red, didn't show 'redes'
After a few searches for the above experiment, it "run out of searches" in other words, I had to reset to see all lessons (had to click a couple of times)
It is accent mark sensitive - if missing it won't find the word
The * search retrieves "all" possible results for the use of a root, ie. 'utiliza*' (verbs + nouns)
I did some searches in French and this is what I found there:

'avancée' retrieves avancée, avancées, avancés.

'caractéristique' retrieves the plural 'caractéristiques'

l'etu(missing the

It doesn't retrieve a string of characters that doesn't show the word completely? I do 'orig' thinking of original.
-- origin , retrieves original; original retrieves origin

~It looks to me that after a few searches the search lesson stops working until you reset.~ Now fixed 🎉 

In regards to what gets returned, lunr does something call term presence and boosting to score documents. I think it would be probably helpful to have the tech team read the lunr documentation on searching for future tests [https://lunrjs.com/guides/searching.html](https://lunrjs.com/guides/searching.html) to help clarify how lunr works. Also would be helpful to a few of us who can explain it to the rest of the team.

jenniferisasi commented:
Thank you Zoe for the amazing work! When would it be a good due date for a second/final round of testing?

ZoeLeBlanc commented:
@jenniferisasi good question and one I don't have a clear answer for 🤔 . I would assume before the next tech meeting but not sure when that's scheduled. Otherwise I'm not sure we have a clear deadline beyond sooner is better than later 🤷‍♀️ 

I need to write up some detailed tech documentation for the search to go on the wiki and then maybe a blog post introducing search so we have time. But @mdlincoln do you have thoughts on timeline at all?

mdlincoln commented:
Can we finish implementing the technical stuff by next Friday? So that you can finish docs & blog post and we can launch by the end of the month?

mdlincoln commented on js/lessonfilter.js:
```diff
@@ -60,51 +127,128 @@ function wireButtons() {
+    $.getJSON(`https://programminghistorian.org/${language}/search.json`).done(response => {
+      corpus = response;
+      // Enable search input and button once data loaded
+      $('#search').attr("placeholder", "Type search terms...");
```
Since we can't (easily) reference the translation snippets YAML from javascript, is there a way you could add a hidden element to lesson-index.html where we CAN load in `{{ site.data.snippets.type-search-terms[page.lang] }}` and then grab its value to insert as the placeholder here? Or some other way so that we don't have to hardcode display language in the JS?

ZoeLeBlanc replied:
That sounds like a great solution! I'll pull down your updates and try it out :)

mdlincoln replied:
(I'm 95% sure this is used on some other part of the site filter JS haha and I'm just lifting the idea)

ZoeLeBlanc replied:
Lolz... I ended up just doing it through `lesson-index.html` and `style.css`, where we just hide and show from inputs from `lessonfilter`. It seemed less complicated but feel free to do any edits!

mdlincoln committed:
create snippet templates for search terms

mdlincoln commented:
I've created said snippet entries so that the search interface can be translated

Commit:
updated search input to work with snippets

ZoeLeBlanc commented:
Input should now work with snippets (just did some simple css rather than running things from lessonfilter).

Final thing is getting those Spanish and French snippets and then a bit more testing 🎉

Commit:
trying to comment and clean up code

mdlincoln commented:
@programminghistorian/french-team and @programminghistorian/spanish-team can you please add and review the new text snippets for the full text search interface?

ZoeLeBlanc commented:
@mdlincoln should I go ahead and clean up the code a bit too? There's some redundant logic now that all the filtering is happening through search. Also wondering if I should be a bit more consistent in using functions vs fat arrows, or var vs let & const... probably overthinking, but let me know if you have any thoughts. Thanks!

spapastamkou committed:
FR snippets

spapastamkou reviewed (commented):
If user wants to enter more than one search terms, do these terms have to be separated in a specific way, and if yes, should we add this in the snippet? Thanks (great work by the way!)

mdlincoln commented:
>should I go ahead and clean up the code a bit too? There's some redundant logic now that all the filtering is happening through search

yeah, fire away. I have no opinion on JS dialects 🙃 it would be great if it were consistent so that future ppl don't get too confused trying to sort through it all.

Commit:
cleaned up code

Commit:
Merge branch 'search-lunr-1018' of github.com:programminghistorian/jekyll into search-lunr-1018

Commit:
ordered lessons by search score

ZoeLeBlanc commented:
Alright here's an updated list of what the latest changes include:
- commented code
- appending the search term to the URI so you can use the back button if you leave the page and still reload your search results
- standardizing code so that we use `function` without fat arrows for functions and event handlers (mostly for binding of `this`), fat arrows for any complex data filtering, and `let` and `const` for all variables.
- DRYing the filter click code now that the filter logic actually happens in search
- Now dynamically updating the number of lessons shown next to `Filtering:`
- Ordering results for search based on lunr search score (included score as a field in featureList)

Some things to consider:
- Now that we are actually using the class `abstract` our abstracts are actually smaller than on the current version of the programming historian. Do we want this change or should I get rid of the styling (since we never used it)?
- To your question @spapastamkou , when you type in multiple search terms Lunr automatically uses an `OR` search so that you would get lessons that contain both or one of the terms. If you want to specify that you want lessons with both you need to add `+` before each term (i.e. `+Twitter +Network`). You can also specify that you don't want certain terms using the `-` before the term (i.e. `+Twitter -Network`). Lunr is super powerful for searching, so not only can you do wildcard searches, but you can also do fuzzy matching (i.e. `TEXT~1` this will pull up words like next or test because it's getting all words that match at least three of these letters) and boosts which will weight certain terms higher (i.e. `Twitter^10 Network`). More details are available here [https://lunrjs.com/guides/searching.html](https://lunrjs.com/guides/searching.html). Do we want to include this information somewhere so that users have some guidelines on how to search?
- When you use the back button and load in a search, it takes awhile for the results to update. Should we include a loader graphic to indicate to users that the results are updating?
- Now that we are ordering search results by lunr, should we indicate this to users or is just assumed?

Let me know if anyone finds any bugs or has any thoughts about these issues. Thanks 🙏

mdlincoln committed:
Merge branch 'gh-pages' into search-lunr-1018

mdlincoln commented:
>Now that we are actually using the class abstract our abstracts are actually smaller than on the current version of the programming historian. Do we want this change or should I get rid of the styling (since we never used it)?

I'd say get rid of it so that the style remains consistent

Commit:
removed abstract styling from style.css

Commit:
Merge branch 'search-lunr-1018' of github.com:programminghistorian/jekyll into search-lunr-1018

mdlincoln commented:
Let's get this wrapped up folks.

@programminghistorian/spanish-team @rivaquiroga we still need translations of the new text snippets

re: documenting how search works, I don't think we should reproduce Lunr's existing docs. I would say, add a small (?) icon that links to their docuentation on wildcards etc for the very tiny portion of users who 1) would even read documentation and 2) badly need to do wildcard searching across such a tiny dataset of ours

I've not noticed the loading problem but maybe I'm not reproducing it correctly?

I believe the implicit reordering by text relevance is expected and welcome and doesn't need additional callouts.

drjwbaker commented:
@programminghistorian/technical-team I've had a play, and I'm wildly impressed. Thanks so much for your hard work on this.

An observation, though, on @mdlincoln's comment about ordering. When I search for - say - `lingustics` I get the output below (url: https://deploy-preview-1726--ph-preview.netlify.app/en/lessons/?search=linguistics&sortType=date&sortOrder=desc) which implies to me that the ordering is by date.

![Screenshot 2020-05-13 at 17 38 23](https://user-images.githubusercontent.com/4965300/81840353-c859df00-9540-11ea-9aea-4233c8cf4644.png)

I then hit date and I get a different ordering.

![Screenshot 2020-05-13 at 17 40 42](https://user-images.githubusercontent.com/4965300/81840436-e7587100-9540-11ea-9d14-09beb659b78a.png)

I then hit date again and get another different ordering.

![Screenshot 2020-05-13 at 17 46 40](https://user-images.githubusercontent.com/4965300/81841039-bb89bb00-9541-11ea-805c-be251fb59e52.png)

What I take from this is that the first search ordering (per https://github.com/programminghistorian/jekyll/pull/1726#issuecomment-626358008) is by relevance. But that isn't clear to me when above the list of results it says `FILTERING RESULTS: (9) DATE ↓`. 

Now I don't actually think this matters that much. **Nobody** is going to use our search for reproducible results. It is more a question of whether or not we are happy with the compromise of the display not matching the ordering after a search.

It isn't my place to make that decision (because sub-teams policy). It is a mere observation.

jenniferisasi committed:
Update snippets.yml

translated to ES search snippets

Commit:
added search info and search indicator

Commit:
switched icon to question mark

Commit:
Merge branch 'gh-pages' of github.com:programminghistorian/jekyll into search-lunr-1018

ZoeLeBlanc commented:
Thanks for the feedback @mdlincoln and @drjwbaker! I've incorporated your suggestions, so now when you search it indicates that the ordering is by search with descending values. If we don't like it we can take it out, since it didn't take much to add it in.

I also added the question mark icon with an added pop out div when clicked, explaining how the search works. Considering all the questions we've had from the tech team, I don't think it's a horrible idea to include some explanation of how to search, along with the link to lunr docs. But let me know if you think it's overkill. It would require adding it to our snippets as well, so might be best to get rid of it.

In terms of the loading issue, I would recommend doing a search and then going to google or some site, and then doing the back button. That should automatically reload in the search queries, which usually takes a second (though maybe your connection is very fast), which means we display default results until the search loads. It's small potatoes but we could easily have a loader if we wanted to. Just let me know!

Also thanks for the translations @jenniferisasi and @rivaquiroga!

Think we're closing in on the final few tweaks 🎉

mdlincoln commented:
Thanks for that careful observation @drjwbaker and super quick work @ZoeLeBlanc ! If the loader is truly small potatoes, then sure put it in! (just make sure it uses shapes/symbols and not something that needs yet more translation)

After you do that, I'll go through the code once more asking for any further clarifications or comments, and then I think we may be ready to merge :)

drjwbaker commented:
Thanks for making the tweaks. The question mark pop out isn't working for me on Firefox (it is on Chrome).

![Screenshot 2020-05-14 at 09 43 50](https://user-images.githubusercontent.com/4965300/81913220-9049af00-95c7-11ea-97a1-72307c60715c.png)

Commit:
fixed search info for firefox and added loader

Commit:
pulled from gh-pages

ZoeLeBlanc commented:
Thanks for catching that @drjwbaker! It's now fixed for firefox. I haven't tested on IE at all so if someone is able to do that it would be helpful (unless we don't support it anyways).

@mdlincoln I added in a loader but it's a bit janky. I'm not an svg animation person, so I'm sure there's a way to get it smoother but it's beyond me for the moment. If anyone wants to try it out, try this url `http://127.0.0.1:4000/en/lessons/?search=twitter+-NETWORK+%2BDATA&topic=python` and let me know if you think we should keep it or toss it. Thanks!!

drjwbaker commented:
Confirm that Firefox now works.

`http://127.0.0.1:4000/en/lessons/?search=twitter+-NETWORK+%2BDATA&topic=python` not loading for me.

ZoeLeBlanc commented:
@drjwbaker hmmm that may be that you're running off a different local host. You could try just typing into the search bar `TWITTER -NETWORK +DATA` and then selecting topic python. Leave the page (to twitter, google, etc... ) and then try the back arrow button to reload search and filter on load. Let me know if that still doesn't work. Thanks!

drjwbaker commented:
Confirm the filters stay on when I get to something like `https://deploy-preview-1726--ph-preview.netlify.app/en/lessons/?search=TWITTER+-NETWORK+%2BDATA&topic=python`, leave the page, then use back arrow.

Great job.

ZoeLeBlanc commented:
Thanks @drjwbaker! Did you see the loading circles by any chance? Also did you like them or not 😅 ? Trying to decide if we want to keep the circles, since it takes a few seconds when loading search results from the URL.

drjwbaker commented:
I did see them! They did a little dance and then let me see the page. It was nice.

Commit:
cleaning up code and adding comments

Commit:
Merge branch 'gh-pages' of github.com:programminghistorian/jekyll into search-lunr-1018

mdlincoln commented:
Thanks Zoe I see your commits - hoping to peer through them this weekend and then we can launch 🎷

mdlincoln commented on js/lessonfilter.js:
```diff
@@ -1,18 +1,22 @@
+  /* Function to update lesson-list using featureList and sort direction
+    - uses URI to generate sort directions
+  */
   console.log("applying sort from URI");
```
Remove the console log functions now that we're ready to push this

ZoeLeBlanc replied:
These were there originally (i.e. before I started working on search) but happy to remove them.

ZoeLeBlanc replied:
Should I also remove where we console.log the new URI?

mdlincoln commented on js/lessonfilter.js:
```diff
@@ -47,74 +51,287 @@ function applySortFromURI(uri, featureList) {
 
   // Performm new sort
```
typo

ZoeLeBlanc replied:
Also there originally but happy to clean it up

mdlincoln reviewed (changes requested):
Could we move the lunr stemming code into a `js/vendor` subdirectory just to make clear that it's drop-in code that didn't need to be customized for our site?

Thinking ahead, how difficult will it e to add in support for Portuguese stemming? Can we make sure that process is documented in the code comments or elsewhere?

mdlincoln commented on js/lessonfilter.js:
```diff
@@ -47,74 +51,287 @@ function applySortFromURI(uri, featureList) {
+  // Create html to show search results using html mark
+  let elements = []
+  docs.map((doc) => {
+    let elementName = '/' + doc.url.split('/').slice(3).join('/')
```
😬  Although we're probably not planning on changing our URLs anytime soon, perhaps add a comment here that this code is based on finding the appropriate relative path for a lesson from the absolute path that the search JSON is returning? (that IS what this is trying to do, yes?)

ZoeLeBlanc replied:
Pretty much! In lesson-describe, we're using the lesson url as a unique identifier but that only produces relative urls (/language/lessons/lesson-name). If we wanted it to be url syntax independent, we could add unique ids to all lessons and use that in generating the corpus and the html. But obviously that would be quite a bit of added grunt work. I'll add a comment here though explaining what's going on.

ZoeLeBlanc replied:
Although I wonder if we could just use the slug for each lesson? That would be url independent... 🤔

ZoeLeBlanc replied:
So slug definitely works instead of using absolute url, but we would need to change the search.json and then the search-index repo to test it. Should I push up a separate PR for the search.json so that we can test all the changes using the slug as id, rather than having to use the absolute url?

ZoeLeBlanc replied:
nvm seems like not all posts have slugs... alright back to just commenting the code

Commit:
cleaning code and changing url to slug

Commit:
Merge branch 'gh-pages' of github.com:programminghistorian/jekyll into search-lunr-1018

Commit:
reverted back to url and added more comments

ZoeLeBlanc commented:
@mdlincoln I like the idea of the js/vendor and happy to document (though it is almost entirely just following the lunr docs). One question I've had is where to document everything? Do you have any suggestions?

Also for the search-info pop out. Are we planning to keep it? If yes, I can add it to our snippets. Finally should I take out all the console.logs or just the one you highlighted. There's quite a few of them so let me know. Thanks!

mdlincoln commented:
Ditch all the console.log()

A high-level overview page similar to the wiki I wrote for the translation links would be very helpful. You don't need to explain every single line of code bc the comments are quite good, but mainly to document the architecture.

And that mini-help text is great. Let's add it to snippets

Commit:
removed all console.logs and added snippet

ZoeLeBlanc commented:
Alrighty! All console.logs are gone and snippet added 🎉 

The example of Twitter and Network in the search info works for Spanish but not French, based on the lessons available. I would recommend for French maybe doing `+Python -Ruby` but it raises the question of whether we may want to standardize examples across languages or just let each team decide which example search terms would be most useful to readers.

I'll start working on the documentation tutorial this week but seems like we're in the final mile 🙌

mdlincoln commented:
@rivaquiroga @programminghistorian/spanish-team 
@spapastamkou @programminghistorian/french-team 

If you can please help translate the final snippet that provides search documentation for users: https://github.com/programminghistorian/jekyll/pull/1726/files#diff-a407e023d60dfefccac66d5a7959003cR375-R384

rivaquiroga committed:
ES translation of search documentation snippet

spapastamkou committed:
FR translation search documentation snippet

mdlincoln committed:
use md shortcut for URLs

mdlincoln reviewed (changes requested):
@ZoeLeBlanc please put the boilerplate JS code into a /vendor directory as discussed above

Commit:
moved js files to vendor and only loaded file for language

ZoeLeBlanc commented:
@mdlincoln thanks for catching that! Just moved the files to vendor directory and updated the lesson index to only load appropriate stemmer for each language (rather than loading them all on each page).

mdlincoln commented:
ohhh good catch, I didn't even think to check it was loading all that extra JS!

ZoeLeBlanc commented:
hmm weirdly isn't showing that I created vendor and moved the files 🤔 Should I try pushing up again??

mdlincoln commented:
oh shoot - at one point I think `vendor/` was added to our gitignore b/c of the way that ruby gems are installed locally at `/vendor/bundle/ruby`

It ought to be `/vendor` in gitignore so that it doesn't ignore all directories that have the name `vendor` in them, just the one at the root

Commit:
adding the vendor files again

ZoeLeBlanc commented:
ahh that makes more sense! I'll go ahead and fix that in gitignore, and then push up again

Commit:
readding js vendor and updatinggitignore
